Failures in com.hazelcast.jet.impl.processor.AsyncTransformUsingServiceP_IntegrationTest

_master_ (commit 55958e30a94d419f740be754876f94fc8dc58f1f)

12 from 14 tests failed in `AsyncTransformUsingServiceP_IntegrationTest` with similar cause.

Failed on IBM JDK8: http://jenkins.hazelcast.com/job/jet-oss-master-ibm-jdk8/249/testReport/com.hazelcast.jet.impl.processor/AsyncTransformUsingServiceP_IntegrationTest/

Stacktrace:
```
com.hazelcast.core.HazelcastInstanceNotActiveException: HazelcastInstance[[127.0.0.1]:5701] is not active!
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getService(ProxyRegistry.java:73)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.<init>(ProxyRegistry.java:55)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.lambda$new$0(ProxyServiceImpl.java:83)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl$$Lambda$510.00000000942A7350.createNew(Unknown Source)
	at com.hazelcast.internal.util.ConcurrencyUtil.getOrPutIfAbsent(ConcurrencyUtil.java:140)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getOrCreateRegistry(ProxyServiceImpl.java:143)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:151)
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.getDistributedObject(HazelcastInstanceImpl.java:353)
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.getMap(HazelcastInstanceImpl.java:172)
	at com.hazelcast.jet.impl.AbstractJetInstance.getMap(AbstractJetInstance.java:130)
	at com.hazelcast.jet.impl.processor.AsyncTransformUsingServiceP_IntegrationTest.before(AsyncTransformUsingServiceP_IntegrationTest.java:111)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:388)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:62)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:50)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:29)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:818)
```

Standard output:
```
Hiccups measured while running test 'test_pipelineApi_mapNotPartitioned[ordered=true](com.hazelcast.jet.impl.processor.AsyncTransformUsingServiceP_IntegrationTest):'
01:57:55, accumulated pauses: 389 ms, max pause: 93 ms, pauses over 1000 ms: 0
```

This is a summary of all the errors in this test class:

```
03:58:01 [ERROR] test_pipelineApi_mapNotPartitioned[ordered=true] HazelcastInstance[[127.0.0.1]:5701] is not active!
03:58:01 [ERROR] test_pipelineApi_mapNotPartitioned[ordered=true] java.lang.NullPointerException: Target cannot be null!

03:58:01 [ERROR] test_pipelineApi_mapPartitioned[ordered=true] HazelcastInstance[[127.0.0.1]:5701] is not active!
03:58:01 [ERROR] test_pipelineApi_mapPartitioned[ordered=true] java.lang.NullPointerException: Target cannot be null!

03:58:01 [ERROR] test_pipelineApi_mapPartitioned[ordered=false] HazelcastInstance[[127.0.0.1]:5701] is not active!
03:58:01 [ERROR] test_pipelineApi_mapPartitioned[ordered=false] java.lang.NullPointerException: Target cannot be null!

03:58:01 [ERROR] stressTest_withRestart_graceful[ordered=false] HazelcastInstance[[127.0.0.1]:5701] is not active!
03:58:01 [ERROR] stressTest_withRestart_graceful[ordered=false] java.lang.NullPointerException: Target cannot be null!

03:58:01 [ERROR] stressTest_noRestart[ordered=false] HazelcastInstance[[127.0.0.1]:5701] is not active!
03:58:01 [ERROR] stressTest_noRestart[ordered=false] java.lang.NullPointerException: Target cannot be null!

03:58:01 [ERROR] test_pipelineApi_mapNotPartitioned[ordered=false] HazelcastInstance[[127.0.0.1]:5701] is not active!
03:58:01 [ERROR] test_pipelineApi_mapNotPartitioned[ordered=false] java.lang.NullPointerException: Target cannot be null!
```

So, two test methods passed (stress tests), taking 17 and 15 seconds to complete. Then the cluster dissolved and everything else failed due to that. This test class uses the serial runner and class-global cluster initialization.

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetInstanceImpl.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetInstanceImpl.java
@@ -18,6 +18,7 @@ package com.hazelcast.jet.impl;
 
 import com.hazelcast.cluster.Address;
 import com.hazelcast.instance.impl.HazelcastInstanceImpl;
+import com.hazelcast.internal.util.Preconditions;
 import com.hazelcast.jet.Job;
 import com.hazelcast.jet.config.JetConfig;
 import com.hazelcast.jet.config.JobConfig;
@@ -57,7 +58,7 @@ public class JetInstanceImpl extends AbstractJetInstance {
 
     @Nonnull @Override
     public List<Job> getJobs() {
-        Address masterAddress = nodeEngine.getMasterAddress();
+        Address masterAddress = getMasterAddress();
         Future<List<Long>> future = nodeEngine
                 .getOperationService()
                 .createInvocationBuilder(JetService.SERVICE_NAME, new GetJobIdsOperation(), masterAddress)
@@ -75,7 +76,7 @@ public class JetInstanceImpl extends AbstractJetInstance {
 
     @Override
     public List<Long> getJobIdsByName(String name) {
-        Address masterAddress = nodeEngine.getMasterAddress();
+        Address masterAddress = getMasterAddress();
         Future<List<Long>> future = nodeEngine
                 .getOperationService()
                 .createInvocationBuilder(JetService.SERVICE_NAME, new GetJobIdsByNameOperation(name), masterAddress)
@@ -86,6 +87,11 @@ public class JetInstanceImpl extends AbstractJetInstance {
         } catch (Throwable t) {
             throw rethrow(t);
         }
+    }
+
+    @Nonnull
+    private Address getMasterAddress() {
+        return Preconditions.checkNotNull(nodeEngine.getMasterAddress(), "Cluster has not elected a master");
     }
 
     @Override

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JetTestSupport.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JetTestSupport.java
@@ -327,13 +327,16 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
             logger.warning("Failed to cancel the job and it is " + status + ", retrying. Failure: " + cancellationFailure,
                     cancellationFailure);
         }
-
         // if we got here, 10 attempts to cancel the job have failed. Cluster is in bad shape probably, shut it down
-        logger.warning(numAttempts + " attempts to cancel the job failed"
-                + (instancesToShutDown.length > 0 ? ", shutting the cluster down" : ""));
-        for (JetInstance instance : instancesToShutDown) {
-            instance.getHazelcastInstance().getLifecycleService().terminate();
+        try {
+            for (JetInstance instance : instancesToShutDown) {
+                instance.getHazelcastInstance().getLifecycleService().terminate();
+            }
+        } catch (Exception e) {
+            // ignore, proceed to throwing RuntimeException
         }
+        throw new RuntimeException(numAttempts + " attempts to cancel the job failed" +
+                (instancesToShutDown.length > 0 ? ", shut down the cluster" : ""));
     }
 
     /**

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceP_IntegrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceP_IntegrationTest.java
@@ -26,7 +26,6 @@ import com.hazelcast.jet.config.EdgeConfig;
 import com.hazelcast.jet.config.JetConfig;
 import com.hazelcast.jet.config.JobConfig;
 import com.hazelcast.jet.core.DAG;
-import com.hazelcast.jet.core.JobStatus;
 import com.hazelcast.jet.core.ProcessorSupplier;
 import com.hazelcast.jet.core.Vertex;
 import com.hazelcast.jet.core.WatermarkPolicy;
@@ -62,7 +61,6 @@ import static com.hazelcast.jet.Traversers.traverseItems;
 import static com.hazelcast.jet.config.ProcessingGuarantee.EXACTLY_ONCE;
 import static com.hazelcast.jet.core.Edge.between;
 import static com.hazelcast.jet.core.EventTimePolicy.eventTimePolicy;
-import static com.hazelcast.jet.core.JobStatus.COMPLETED;
 import static com.hazelcast.jet.core.JobStatus.RUNNING;
 import static com.hazelcast.jet.core.TestUtil.throttle;
 import static com.hazelcast.jet.core.processor.SourceProcessors.streamMapP;
@@ -73,7 +71,6 @@ import static java.util.Arrays.asList;
 import static java.util.stream.Collectors.joining;
 import static java.util.stream.Collectors.toMap;
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
 
 @RunWith(Parameterized.class)
 @Parameterized.UseParametersRunnerFactory(HazelcastSerialParametersRunnerFactory.class)

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceP_IntegrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceP_IntegrationTest.java
@@ -112,7 +112,6 @@ public class AsyncTransformUsingServiceP_IntegrationTest extends SimpleTestInClu
         journaledMap.putAll(IntStream.range(0, NUM_ITEMS).boxed().collect(toMap(i -> i, i -> i)));
         sinkList = instance().getList(randomMapName("sinkList"));
         jobConfig = new JobConfig().setProcessingGuarantee(EXACTLY_ONCE).setSnapshotIntervalMillis(0);
-
         serviceFactory = sharedService(pctx -> Executors.newFixedThreadPool(8), ExecutorService::shutdown);
     }
 
@@ -165,17 +164,9 @@ public class AsyncTransformUsingServiceP_IntegrationTest extends SimpleTestInClu
 
         Job job = instance().newJob(dag, jobConfig);
         for (int i = 0; restart && i < 5; i++) {
-            assertTrueEventually(() -> {
-                JobStatus status = job.getStatus();
-                assertTrue("status=" + status, status == RUNNING || status == COMPLETED);
-            });
+            assertJobStatusEventually(job, RUNNING);
             sleepMillis(100);
-            try {
-                job.restart();
-            } catch (IllegalStateException e) {
-                assertTrue(e.toString(), e.getMessage().startsWith("Cannot RESTART_GRACEFUL"));
-                break;
-            }
+            job.restart();
         }
         assertResultEventually(i -> Stream.of(i + "-1", i + "-2", i + "-3", i + "-4", i + "-5"), numItems);
     }
@@ -185,8 +176,7 @@ public class AsyncTransformUsingServiceP_IntegrationTest extends SimpleTestInClu
         Pipeline p = Pipeline.create();
         p.readFrom(Sources.mapJournal(journaledMap, START_FROM_OLDEST, EventJournalMapEvent::getNewValue, alwaysTrue()))
          .withoutTimestamps()
-         .mapUsingServiceAsync(serviceFactory,
-                 transformNotPartitionedFn(i -> i + "-1"))
+         .mapUsingServiceAsync(serviceFactory, transformNotPartitionedFn(i -> i + "-1"))
          .setLocalParallelism(2)
          .writeTo(Sinks.list(sinkList));
 
@@ -200,8 +190,7 @@ public class AsyncTransformUsingServiceP_IntegrationTest extends SimpleTestInClu
         p.readFrom(Sources.mapJournal(journaledMap, START_FROM_OLDEST, EventJournalMapEvent::getNewValue, alwaysTrue()))
          .withoutTimestamps()
          .groupingKey(i -> i % 10)
-         .mapUsingServiceAsync(serviceFactory,
-                 transformPartitionedFn(i -> i + "-1"))
+         .mapUsingServiceAsync(serviceFactory, transformPartitionedFn(i -> i + "-1"))
          .setLocalParallelism(2)
          .writeTo(Sinks.list(sinkList));
 

